### DISCUSSION TITLE
RAM optimizations

### DISCUSSION THREAD

**Discussion by jrahlf** [bitbank2/AnimatedGIF#92] [General]
Can we add a section to the wiki regarding RAM optimization?
The following things are unclear:
- `MAX_CODE_SIZE` how big does it need to be for a given screen size and palette? What does this actually do?
- `MAX_COLORS`: what happens if we set a smaller value than an image actually has? E.g. `MAX_COLORS` 16 but image has 255 colors

Optimization idea:
- `unsigned short pPalette[(MAX_COLORS * 3)/2]`: I don't know why this is  `* 3 / 2`. Shouldn't `MAX_COLORS * 2` be sufficient for 16-bit RGB565?

**Comment by bitbank2** [comment]
The changes you're requesting are a very narrow use case and would cause more confusion than help to most users. I suggest you create a fork of the repo for you to experiment. If you really need to trim down this memory, then the target platform probably should work with GIF images.

**Reply by jrahlf** [reply]
Can you still answer the questions?

**Reply by tobozo** [reply]
> Shouldn't MAX_COLORS * 2 be sufficient for 16-bit RGB565?

probably in a RGB565-only world, but the buffer is shared with RGB888 so `*3` it is.

```cpp
  unsigned short pPalette[(MAX_COLORS * 3)/2]; // can hold RGB565 or RGB888 - set in begin()
```

unless you're on Arduino Nano, saving 1/3rd of 768 bytes isn't really an optimization anyway

> MAX_CODE_SIZE how big does it need to be for a given screen size and palette? What does this actually do?

it affects the file buffer size : `FILE_BUF_SIZE (1<<MAX_CODE_SIZE)`, a lower value will cause memory fragmentation, and a higher value may cause stuttering

`MAX_CODE_SIZE` is a power of 2, i.e. if MAX_CODE_SIZE=12, then buffer size = 4096

> MAX_COLORS: what happens if we set a smaller value than an image actually has? E.g. MAX_COLORS 16 but image has 255 colors

by all means, please tell us what happens!

**Reply by jrahlf** [reply]
> > Shouldn't MAX_COLORS * 2 be sufficient for 16-bit RGB565?
> 
> probably in a RGB565-only world, but the buffer is shared with RGB888 so `*3` it is.

I would understand `MAX_COLORS * 2` for RGB565 and `MAX_COLORS * 3` for RGB888, but why is it divided by 2?

**Reply by jrahlf** [reply]
Ok I oversaw the type is 16-bit short, I assumed it was bytes

**Comment by bitbank2** [comment]
For normal GIF images, MAX_CODE_SIZE is a fixed value. 12-bit codes require a table with 4096 entires. If you make GIFs with smaller LZW tables, you could change that value, but it wouldn't be able to decode standard files. The color palette is a tiny amount of RAM compared to the other buffers. Saving a couple of hundred bytes mostly is just a rounding error.

**Reply by jrahlf** [reply]
Okay, so I understand that `usGIFTable` and `ucGIFPixels` need to be 4096 and 8192 bytes to decode standard GIFs.

However, the `FILE_BUF_SIZE` determines how many bytes of encoded GIF data can read at a time (either from RAM or file) and then used as the working data. I reckon if reading from RAM, lowering this value should not be too detrimental. Also, maybe one wants to increase this value.
I think the `FILE_BUF_SIZE` can thus be made arbitrary in size (with a certain minimum value)?

PS:
in this line https://github.com/bitbank2/AnimatedGIF/blob/accf23cdccbc4e92fcd4a7e21bb429e8c52f78a4/src/gif.inl#L1408 it seems to assume a fixed size of 4096 currently for the `usGIFTable`.

**Comment by bitbank2** [comment]
Can you state why your are interested in shaving off a few bytes here and there? Do you have a very constrained system that doesn't have enough RAM? Maybe GIF isn't the right format to use for your task.

**Reply by jrahlf** [reply]
This library is supposed to be optimized for RAM usage, so why "waste" a few hundreds or even single digit kilobyte of RAM?
If you have a larger project consisting of many modules, it adds up.
In my case, the available RAM is trending <20kB on an esp32s3 so I am happy if I can free 1kB somewhere.

**Comment by bitbank2** [comment]
I understand your use case, but it's too specific to you to potentially disturb things for the average user. I think it would be best for you to make a fork and then you're free to modify it any way you want.

**Comment by jrahlf** [comment]
There is another optimization to be made:
Opening a GIF from Flash/RAM does not require the intermediate `ucFileBuf` at all. They already require the source to be valid for the entire duration of the GIF rendering. So one could save the 4kB (current buffer size) and also omit the memcpys.

**Reply by bitbank2** [reply]
This is incorrect. Part of the optimization of my code is being able to quickly read the compressed code words without checking for chunk boundaries in the inner decode loop. The way it works is that I de-chunk the GIF file data into the ucFileBuf before running the decode loop. Please take the time to actually look through the code before making assumptions. There are plenty of comments which explain the hows and whys along the way.
